server/lib/src/plugins/namehistory.rs

```diff
@@ -0,0 +1,351 @@
+    EntryRefreshNew, EntrySealedCommitted, EntrySealedNew,
+};
+use crate::event::{CreateEvent, DeleteEvent, ModifyEvent};
+use crate::filter::f_eq;
+use crate::modify::{Modify, ModifyList, ModifyValid};
+use crate::plugins::Plugin;
+use crate::prelude::*;
+use crate::prelude::{
+    BatchModifyEvent, QueryServerReadTransaction, QueryServerWriteTransaction, ValueSet,
+    PVCLASS_SYSTEM_CONFIG,
+};
+use crate::schema::SchemaTransaction;
+use crate::server::QueryServerTransaction;
+use crate::value;
+use crate::value::PartialValue;
+use crate::value::Value::Utf8;
+use crate::valueset::ValueSetUtf8;
+
+pub struct NameHistory {}
+
+#[derive(Eq, Hash, PartialEq, Clone)]
+enum name_type {
+    Name,
+    LegalName,
+    DisplayName,
+}
+
+impl ToString for name_type {
+    fn to_string(&self) -> String {
+        match self {
+            name_type::Name => "name",
+            name_type::LegalName => "legalname",
+            name_type::DisplayName => "displayname",
+        }
+        .to_string()
+    }
+}
+
+impl FromStr for name_type {
+    type Err = &'static str;
+
+    fn from_str(s: &str) -> Result<Self, Self::Err> {
+        return match s {
+            "name" => Ok(name_type::Name),
+            "legalname" => Ok(name_type::LegalName),
+            "displayname" => Ok(name_type::DisplayName),
+            _ => Err("must be either: name, legalname, or displayname"),
+        };
+    }
+}
+
+lazy_static! {
+    // the naming can be confusing, here it means: 'the set of the uuids of the entries we want to target'
+    static ref entries_uuids_to_update: HashSet<PartialValue> = {
+        let mut set = HashSet::new();
+        set.insert(PartialValue::new_iutf8("account"));
+        set
+    };
+}
+
+impl NameHistory {
+    //this function returns the entries to update together with their original index they had
+    //in the cand vector. This is useful as original index is often needed to read the
+    //pre_cand vector which shares the same entries and thus the same indexes with the cand vector
+    fn entries_to_update<VALID, STATE>(
+        cand: &mut Vec<Entry<VALID, STATE>>,
+    ) -> Vec<(usize, &mut Entry<VALID, STATE>)> {
+        let mut target_cand = cand
+            .iter_mut()
+            .enumerate()
+            .filter_map(|(i, elm)| {
+                let is_target_entry = entries_uuids_to_update.iter().fold(true, |acc, current| {
+                    elm.attribute_equality("class", current) && acc
+                });
+                return if is_target_entry {
+                    Some((i, elm))
+                } else {
+                    None
+                };
+            })
+            .collect::<Vec<(usize, &mut Entry<VALID, STATE>)>>();
+        return target_cand;
+    }
+
+    fn updated_names(modification: &ModifyList<ModifyValid>) -> HashSet<name_type> {
+        let mut modified_names: HashSet<name_type> = HashSet::new();
```

Comment:
There may be an easier way to do this that avoids allocating a hashset.

Also worth noting that while hashsets are cool and all, they are *expensive* to allocate and free. It's better to simply have a btreeset for small sets. 

BUT - IMO there is likely a way to achieve this without needing a hashset at all :)

Reply:
I nearly rewrote everything following what you wrote above so the hashset just disappeared :p